File: c++.template.f
//1.2.6模板的局限性
//模板不是万能的，有些特定数据类型，需要具体化方式做特殊实现
//对比两个数据是否相等函数
#include<iostream>
using namespace std;
class Person{
public:
	Person(string name, int age){
		m_Name = name;
		m_Age = age;
	}
	//姓名
	string m_Name;
	//年龄
	int m_Age;
};
template<class T>
bool myCompare(T &a, T&b){
	if (a == b){
		return true;
	}
	else{
		return false;
	}
}
//利用具体化Person的版本实现代码，具体优化先调用
template<>bool myCompare(Person &p1, Person &p2){
	if (p1.m_Name == p2.m_Name&&p1.m_Age == p2.m_Age){
		return true;
	}
	else{
		return false;
	}
}
void test01(){
	int a = 10;
	int b = 20;
	bool ret = myCompare(a, b);
	if (ret){
		cout << "a==b" << endl;
	}
	else{
		cout << "a!=b" << endl;
	}
}
void test02(){
	Person p1("Tom", 10);
	Person p2("Tom", 10);
	bool ret = myCompare(p1, p2);
	if (ret){
		cout << "p1==p2" << endl;
	}
	else{
		cout << "p1!=p2" << endl;
	}
}
int main(){
	//test01();
	test02();
	system("pause");
	return 0;
}

File: c++.template.m
1.3.8 类模板与友元
全局函数类内实现-直接在类内声明即可
全局函数类外实现-需要提前让编译器知道全局函数的存在
#include<iostream>
using namespace std;
#include<string>
//通过全局函数打印Person信息
//提前让编译器知道Person类存在
template<class T1, class T2>
class Person;
//类外实现
template<class T1, class T2>
void printPerson2(Person<T1, T2>p){
	
	cout << "类外实现姓名：" << p.m_Name << " 年龄：" << p.m_Age << endl;
}
template<class T1, class T2>
class Person{
public:
	Person(T1 name, T2 age){
		this->m_Name = name;
		this->m_Age = age;
	}
	//全局函数 类内实现
	friend void printPerson(Person<T1, T2>p){
		cout << "姓名：" << p.m_Name << " 年龄：" << p.m_Age << endl;
	}
	//加空模板参数列表
	//如果全局函数是类外实现，需要让编译器提前知道这个函数的存在
	friend void printPerson2<>(Person<T1, T2>p);
		
private:
	T1 m_Name;
	T2 m_Age;
};
////类外实现
//template<class T1, class T2>
//void printPerson2(Person<T1, T2>p){
//	cout << "类外实现姓名：" << p.m_Name << " 年龄：" << p.m_Age << endl;
//}
	
//1、全局函数在类内实现
void test01(){
	Person<string, int>p("Tom", 20);
	printPerson(p);
}
//2、全局函数在类外实现
void test02(){
	Person<string, int>p("Jey", 20);
	printPerson2(p);
}

int main(){
	//test01();
	test02();
	system("pause");
	return 0;
}
